### ISSUE TITLE
How to test react-pixi components

### ISSUE THREAD

**Issue by bichotll** [inlet/react-pixi#38]
### Description

It would be good if we can also write on the documentation about how to test our components.
_(They print on a canvas, not change DOM elements. The changes can not be marked and asserted/checked using anything similar css rules or dom elements properties. Example: Graphics.draw method.)_
It does not seem to be as straight forward as testing normal (dom) components.

Once I know how these can be tested, I can write a documentation page for it.

### My problem

I'm investigating how I should test my components and I'm having a bit of trouble so far.
Once I get it working, I think I'm gonna save an image of the canvas to have something to compare it to in future tests/changes...similar to snapshots.

#### What I've tried

So far I tried using Jest and Enzyme (shallow), with and without wrapping it with <Stage>.
What I'm getting all the time is `Cannot set property 'fillStyle' of null`.

_Edit ()_: Clarification regarding the test of the components.

**Comment by bichotll** [comment]
Alright, for the record, I figured out that I needed `canvas` installed for jsdom.

For more information:
https://github.com/jsdom/jsdom#loading-subresources

Now I'm getting `TypeError: Cannot read property 'alternate' of undefined` coming from `node_modules/@inlet/react-pixi/dist/react-pixi.production.umd.js` whenever I try to Implement the Stage component before my test...
Working on that.

**Comment by bichotll** [comment]
Alright, again, for the record.

## How to test the pixi-react components comparing them to a snapshot

Here is what I did to test how the component is being printed on the canvas.

### My utils

I have installed `canvas` package (npm).

As an addition - I created a `bootstrap.js` file like the one we have in this project just to avoid the nice hello message:
```js
require('pixi.js').utils.skipHello();
``` 
I created a `react-pixi-util.js` file to handle the creation of the canvas, pixi app and the snapshot.
```js
import * as PIXI from 'pixi.js';
import { render } from '@inlet/react-pixi';

const initPixiPlayground = (document, options = {
  width: 300,
  height: 300,
  backgroundColor: 0x10bb99,
  forceCanvas: true,
}) => {
  const canvas = document.createElement('canvas');
  canvas.id = 'canvas';
  document.body.appendChild(canvas);

  return new PIXI.Application({
    ...options,
    view: canvas,
  });
};

const onPostRenderPromise = pixiApp => (
  new Promise((resolve) => {
    pixiApp.renderer.on('postrender', () => {
      resolve();
    });
  })
);

const expectMatchSnapshot = (pixiApp, componentToRender) => {
  render(componentToRender, pixiApp.stage);
  return onPostRenderPromise(pixiApp).then(() => {
    expect(pixiApp.view.toDataURL('image/png')).toMatchSnapshot();
  });
};

export default {
  initPixiPlayground,
  onPostRenderPromise,
  expectMatchSnapshot,
};
```

### Example of test
```js
import React from 'react';
import { render } from '@inlet/react-pixi';
import reactPixiUtil from '[your file location]/react-pixi-util';
import YourPixiComponent from './index';

describe('Feature: <YourPixiComponent />', () => {
  let pixiApp;

  beforeEach(() => {
    pixiApp = reactPixiUtil.initPixiPlayground(document);
  });

  describe('snapshots', () => {
    it('non active', () => (
      reactPixiUtil.expectMatchSnapshot(pixiApp, () => {
        render(<YourPixiComponent active={false} />, pixiApp.stage);
      })
    ));

```

## What we can do next

I'd like to discuss with you @inlet to know what is the best way we can document that and maybe even implement basic tests here using my debugging/research? 👍

**Comment by inlet** [comment]
Hi @bichotll,

First, thanks for taking the time and to contribute!

I hope I follow you correctly, if not please let me know 😊. 

#### Testing

Have you looked at the test code in this repo?
You can use jest snapshots to compare and use `.getInstance()` to retrieve the PIXI elements for internal works.

I just have'd a quick look at your example test code and you're comparing the output of a png (base64), although this might work I think it's better to compare the React components blueprint (toJSON, describing the components itself including props).

#### Docs

I'd like to keep things simple and the docs (for this repo) is meant to demo how the API works and how to use it in your project, see https://reactpixi.org. Thinks like testing is out of this scope, I presume you can test React-PIXI components the same way as you'd test other React components, in that sense they are the same.

If you'd like to discuss this further, you can always [drop me a mail](mailto:patrick@inlet.nl) 🙌

Thanks! Patrick

**Comment by bichotll** [comment]
Hi @inlet 
There is something I should have clarified at the beginning of the post and I see I barely talked about it...😅pardon! On top of that I was also having trouble testing it.


### Test the changes on the canvas

One of the important points when talking about react pixi components is that they modify a canvas. The canvas is not like the DOM where changes are traceable like css for example.
Let's give an example of a simple button:
- Using the DOM: you could create with a simple `div` element and then assigning it a css property (border: 4px) when hovered. To make sure it works, you need to check the css property border to make sure the button changed its appearance.
- Using the Canvas: let's say you use a Graphics component and you repaint (using draw) the button when hovered. That does not leave any log or property on the canvas that can be checked. The best way you can make sure the button will have the desired appearance will be taking a picture of it and making sure that's what you are expecting.

I worked for a website years ago that worked heavily with canvas (fabric js). The best way we achieved the testing of the canvas changes was taking "screenshots" of the canvas element.

In conclusion, I think this should be documented one way or another. Maybe not explicitly on the main documentation.
I'll email you, let's see what we can get from all this 👍
And I'm closing this issue for now 👍 

Cheers

**Comment by lucas-jones** [comment]
@bichotll This was super useful. Thank you.
